Sources/SourceKitLSP/Swift/Refactoring.swift

```diff
@@ -0,0 +1,124 @@
+  ///
+  /// - Parameters:
+  ///   - title: The title of the refactoring action.
+  ///   - dict: Response dictionary to extract information from.
+  ///   - url: The client URL that triggered the `semantic_refactoring` request.
+  ///   - keys: The sourcekitd key set to use for looking up into `dict`.
+  init?(_ title: String, _ dict: SKDResponseDictionary, _ snapshot: DocumentSnapshot, _ keys: sourcekitd_api_keys) {
+    guard let categorizedEdits: SKDResponseArray = dict[keys.categorizedEdits] else {
+      return nil
+    }
+
+    var refactoringEdits = [RefactoringEdit]()
+
+    categorizedEdits.forEach { _, value in
+      guard let edits: SKDResponseArray = value[keys.edits] else {
+        return false
+      }
+      edits.forEach { _, value in
+        // The LSP is zero based, but semantic_refactoring is one based.
+        guard let startLine: Int = value[keys.line],
+          let startColumn: Int = value[keys.column],
+          let endLine: Int = value[keys.endLine],
+          let endColumn: Int = value[keys.endColumn],
+          let text: String = value[keys.text]
+        else {
+          return true  // continue
+        }
+
+        let startPosition = snapshot.positionOf(
+          zeroBasedLine: startLine - 1,
+          utf8Column: startColumn - 1
+        )
+        let endPosition = snapshot.positionOf(
+          zeroBasedLine: endLine - 1,
+          utf8Column: endColumn - 1
+        )
+        // Snippets are only supported in code completion.
+        // Remove SourceKit placeholders in refactoring actions because they can't be represented in the editor properly.
+        let textWithSnippets = rewriteSourceKitPlaceholders(in: text, clientSupportsSnippets: false)
+        refactoringEdits.append(
+          (
+            startPosition: startPosition, endPosition: endPosition, newText: textWithSnippets,
+            bufferName: value[keys.bufferName]
+          )
+        )
+        return true
+      }
+      return true
+    }
+
+    guard refactoringEdits.isEmpty == false else {
+      return nil
+    }
+
+    self.init(title: title, uri: snapshot.uri, refactoringEdits: refactoringEdits)
+  }
+}
+
+extension SwiftLanguageService {
+  /// Provides detailed information about the result of a specific refactoring operation.
+  ///
+  /// Wraps the information returned by sourcekitd's `semantic_refactoring` request, such as the necessary edits and placeholder locations.
+  ///
+  /// - Parameters:
+  ///   - command: The semantic `RefactorCommand` that triggered this request.
+  ///   - responseType: The response type `T.Type` of the particular command
+  /// - Returns:
+  ///   - The `Refactoring` as `T`
+  func refactoring<T: Refactoring>(_ refactorCommand: any RefactorCommand, _ responseType: T.Type) async throws -> T {
```

Comment:
Instead of explicitly passing in the response type here, what do you think about adding a 

```swift
associatedtype Result: Refactoring
```

to `RefactorCommand` and then changing the function definition to


```suggestion
  func refactoring<T: RefactoringCommand>(_ refactorCommand: T, _ responseType: T.Result) async throws -> T.Result {
```